### ISSUE TITLE
DOCKER_RAMDISK lost effect after upgrade.

### ISSUE THREAD

**Issue by IndeedPlusPlus** [moby/moby#36120]
**Description**
In a system running entirely in memory, after an upgrade from  `17.09.1-ce` to `17.12.0-ce`, docker stopped creating containers, failing with message like `docker: Error response from daemon: OCI runtime create failed: container_linux.go:296: starting container process caused "process_linux.go:398: container init caused \"rootfs_linux.go:107: jailing process inside rootfs caused \\\"pivot_root invalid argument\\\"\"": unknown.`.

The environment variable `DOCKER_RAMDISK` has been always set for both client and server.

**Steps to reproduce the issue:**
1. create a minimal system image to run Docker in a system running in memory.
```bash
# create minimal OS image

docker run -it --name rootfs ubuntu:xenial

# inside container
apt-get update
apt-get install linux-image-generic

# setup networking
cat <<EOF > /etc/systemd/network/20-wired.network
[Match]
Name=en*

[Network]
DHCP=yes

EOF

systemctl enable systemd-networkd systemd-resolved systemd-networkd-wait-online

# install docker
apt-get install apt-transport-https curl
echo 'deb [arch=amd64] https://download.docker.com/linux/ubuntu xenial stable' > /etc/apt/sources.list.d/docker.lis
apt-get update
apt-get install docker-ce

# set a password

passwd

# setup init
ln -s /lib/systemd/systemd /init

exit

# outside container
mkdir ubuntu
docker export rootfs | sudo tar -C ubuntu -xv
cd ubuntu
sudo ln -sf /run/systemd/resolve/resolv.conf ./etc/resolv.conf

# create rootfs
sudo bash -c 'find . | cpio -H newc -o | xz -v --check=crc32 -6 -T0 > /tmp/initrd.cpio.xz'
```
2. Create a virtual machine with direct kernel boot. use `ubuntu/vmlinuz` as kernel and `/tmp/initrd.cpio.xz` as initrd, then boot this virtual machine.

3. inside virtual machine, launch `dockerd` and try to start a container
```bash
systemctl stop docker
export DOCKER_RAMDISK=x
# start dockerd
/usr/bin/dockerd &
# start container
docker run --rm -it busybox
```

**Describe the results you received:**
dockerd output
```
INFO[0102] shim docker-containerd-shim started           address="/containerd-shim/moby/36570cb64fe89bcdbd0e8e582795455afcff6d01adc2c9cf442f8603ee065374/shim.sock" debug=false module="containerd/tasks" pid=2347
INFO[0102] shim reaped id=36570cb64fe89bcdbd0e8e582795455afcff6d01adc2c9cf442f8603ee065374 module="containerd/tasks"
ERRO[2018-01-26T03:53:02.869112354Z] stream copy error: reading from a closed fifo 
ERRO[2018-01-26T03:53:02.997287452Z] 36570cb64fe89bcdbd0e8e582795455afcff6d01adc2c9cf442f8603ee065374 cleanup: failed to delete container from containerd: no such container 
ERRO[2018-01-26T03:53:03.001954258Z] Handler for POST /v1.35/containers/36570cb64fe89bcdbd0e8e582795455afcff6d01adc2c9cf442f8603ee065374/start returned error: OCI runtime create failed: container_linux.go:296: starting container process caused "process_linux.go:398: container init caused \"rootfs_linux.go:107: jailing process inside rootfs caused \\\"pivot_root invalid argument\\\"\"": unknown 
```
docker cli output
```
docker: Error response from daemon: OCI runtime create failed: container_linux.go:296: starting container process caused "process_linux.go:398: container init caused \"rootfs_linux.go:107: jailing process inside rootfs caused \\\"pivot_root invalid argument\\\"\"": unknown.
```

**Describe the results you expected:**
a container started correctly, showing a shell.

**Additional information you deem important (e.g. issue happens only occasionally):**
This issue does not affact versions no later than `17.09.1-ce`.

**Output of `docker version`:**
I tried docker-ce packages from both stable and edge repo. This problem also affects `17.12.0-ce`.
```
Client:
 Version:	18.01.0-ce
 API version:	1.35
 Go version:	go1.9.2
 Git commit:	03596f5
 Built:	Wed Jan 10 20:11:05 2018
 OS/Arch:	linux/amd64
 Experimental:	false
 Orchestrator:	swarm

Server:
 Engine:
  Version:	18.01.0-ce
  API version:	1.35 (minimum version 1.12)
  Go version:	go1.9.2
  Git commit:	03596f5
  Built:	Wed Jan 10 20:09:37 2018
  OS/Arch:	linux/amd64
  Experimental:	false

```

**Output of `docker info`:**

```
Containers: 0
 Running: 0
 Paused: 0
 Stopped: 0
Images: 1
Server Version: 18.01.0-ce
Storage Driver: overlay2
 Backing Filesystem: tmpfs
 Supports d_type: true
 Native Overlay Diff: true
Logging Driver: json-file
Cgroup Driver: cgroupfs
Plugins:
 Volume: local
 Network: bridge host macvlan null overlay
 Log: awslogs fluentd gcplogs gelf journald json-file logentries splunk syslog
Swarm: inactive
Runtimes: runc
Default Runtime: runc
Init Binary: docker-init
containerd version: 89623f28b87a6004d4b785663257362d1658a729
runc version: b2567b37d7b75eb4cf325b77297b140ea686ce8f
init version: 949e6fa
Security Options:
 apparmor
 seccomp
  Profile: default
Kernel Version: 4.4.0-112-generic
Operating System: Ubuntu 16.04.3 LTS
OSType: linux
Architecture: x86_64
CPUs: 8
Total Memory: 7.796GiB
Name: localhost
ID: 3EJN:NZ7J:VC7T:7ESO:WI7I:JI6C:TMZC:C6LB:TSCH:HE56:AXSA:NPID
Docker Root Dir: /var/lib/docker
Debug Mode (client): false
Debug Mode (server): false
Registry: https://index.docker.io/v1/
Labels:
Experimental: false
Insecure Registries:
 127.0.0.0/8
Live Restore Enabled: false

WARNING: No swap limit support
```

**Additional environment details (AWS, VirtualBox, physical, etc.):**
This experiment is conducted within a virtual machine powered by KVM.

**Comment by thaJeztah** [comment]
Thanks for the detailed report; this issue has already been fixed in master through https://github.com/moby/moby/pull/35957 (and will be in the upcoming 18.02 release of docker); I also opened a request internally to have the fix backported to 17.12 if there will be a patch release.

Given that this is already fixed in this repo, I’ll close this issue, but feel free to continue the conversation.

**Comment by IndeedPlusPlus** [comment]
Thank you. Glad to hear that it will be fixed soon.

**Comment by prologic** [comment]
Has this been fixed in 18.03.1-ce ? I get a very similar (*it not identical*) error trying to create a container from a system booted entirely out of ramdisk.

**Comment by prologic** [comment]
Never-mind; false alarm :)

```
# docker run -i -t alpine /bin/sh
/ # cat /etc/alpine-release
3.7.0
/ # #
```

I *forgot* to set `DOCKER_RAMDISK=true` :D
